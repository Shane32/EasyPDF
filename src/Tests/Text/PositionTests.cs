using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
