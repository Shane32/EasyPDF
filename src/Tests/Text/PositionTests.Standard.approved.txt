%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 921/Filter/FlateDecode>>stream
x\x9C\x8D\x97Mo\xDB0\x0C\x86\xEF\xFE\x15>\xB6\x87x\xA2l\xCB\xD6n\xEB\xB0(vY\x97\xDB\xD0\xC3\x90\xA6]\x87\xBA]\xBB\x00\xFB\xFB\x93I\x8A\xA4\x82\x06\x15\xFA\xC5\xC8$\xC5<u\xDEW~n.\xB6M\x1F\xDAi\x0C\xED\xF6\xA6\xF9\xB4m\xBE5\xD0\xBA\xF4\xB5\x81v\xF2\xE9\xDB\xB5\xBB\xA5q\xEDe\xFA\xF9\xDD\xB8\x0E\xDA\x7F\xCD\x8F\xEBt\xFD\xA6q\x98w\xF5\x85\x83\x97\xBB\xD7\xD3\xBCw\xDD\x98\xA2\x85#\xDF\xB1\xF50\xAF\x7F\x86\xF55\x04\xFC\xBBk<\xF4\xDD\x08\xBC\x08\x9A
X\xBF^_\xA3\xAFb\xEE&w\xD8\xE4\x16\xD8\x97\x96q;N\xE7!v\xCD\xF7&7\\d\xB4\x07\\\xE5&K\x93'Z\x97\x9F\xD3\x1B\xDA\xEEVJ\xEF>C\x0B\xE9\xC5\xAD\x02Z\x13\xA7\x0E\xE6v\xBBX\x08nE\xF9\xE3\xEC\xC3\xC3\xFD\xDD\xE3\xB2\x7F<\xBC?\x1F\xDDY\xFBu\x7F{\xD8\x9E\x83wgO\x7F\xCE\xAF\xB7\x97\x04\x9BFH-\x04\x8F3|\xBC\xE7\xB7\x87\x81\x10\xE2e0\xE9@]2$\x181\x8F\xD2\xD7\xDF\xD8\x05\x03\x81D\xCB\xB4?\xA5\xE7Y\x94\x92\x19,\x85\x99\x13V\x12'\x1C\xA5
\x94O\xFDb5\xA9\x8F\xE9\xD5\xFE\xE5\x98S\xBAS\xF38\xFDl\xEF#\x97o$W\xDCIN8\xE5t\xA0.\x99S\xDF\x1BN=p\x17\x0C\x84\x13-\xD3\xA6\xBDr\xC2.\x99\x93\x1D,\x08'\xAC\xE4\xFB\xC9\xD5rJ\xF5\xB5\x90.~\xFE\xDD?\xDC?\xEE\x8F1\x8D\x83L3\x06\x83i\x9C\xF9b \x98x\x19L:P\x97\x8Ci\x04\x83i\x88\x19v\xB4\x98h\x996\x05\xC5\x84]2&;\xD8 \x98\xB0\x920\xE1(U\x98\xD2\x1E\xB1\xAF'\xF5t8<-\xC7\x9C\x92\xC0\xE5q\xA6\xC1p\x9A\x02\xBFC\x0C\x84\x13/\x83I\x07\xEA\x929\x85h8\x85\x89\xBB` \x9Ch\x197\xE5\xF4<\x8Br\xB2\x83y\xE1\x84\x95\xC4\x09Gy\x93\x13\x84\xB9\xF3I\xC4c\xADF\xD1\xA7\xEE\x94JE\xD5\xCChU<f\x11\x8E\x85\x8EG\x15\xF2h\x94<\xAA\x94\xCF\x93\xC15\x8F\xDC\x05\x03\xC1E\xCB\xB8)\xA7\xE7Y\x14\x97\x1DL\xD5\x1C+\x09W\xAC\x92s\x08\xD0\xF9\xBE\x8D}\xADP\x11\xAE\xD7\xA5
\x9C\x91N\xB0\xA2\x0E\x90U\x9D"\x01\x96/\x80-\x01\xEE%\xCA\xEE
iw\xA2\xED\xAE\x14wg\xD4\xDDYyw\x85\xBE\xDB)\x9DQx\xA7\x12Oc\xBDMo\x9C:\x17\xB0K=\xBBS
\x06^\x05\x15\xBC\x95z\xE8\xB3\xD6S\xA4\xF4zU{)\x01\xEE%\xF4\xBC\x15|\xF0Y\xF1)Rz^5?\x97\xC8\\\x86^1\xA5\xEA>\x953\xBD\xBEJ\xF9!\xB8\xAE\x8F\xB8Y\xA5\xAC1\xC0W\x85\x06\x15Z\x18\xAC\x05\xC0\x90=\x80"\xC57\xA8\x0BH\x09p/\xC17@q\xAE\x88\xF2\xAF(\xAC\x80/\xF0\xF6\xC6\x0C\xA8\x97\xE0+\xA6\x1C\xCC\xF1B\x81\xC6z\x1B\x9FO\xE7\xB0\xF5\x1E\x84Z\xA9\xBB\xBA\xBF\xFBu\xFA<\x16T\x80!Xk\x80\x90\xBD\x81"\xE5\x17\xD4\x1D\xA4\x04\xB8\x97\x1E\xCB\xACA\xAC\x9F\x18\xEE5\x16\x16\xC1\x17\xF8hfL\x82z\x09\xBFbJ\xF5\x09*g~\xA1\xCE)\xD2D\xD0\xA7\xEC\xB1V\xFB\x90\xDF\x09\xE9\x9B\x9D95Z\xA7\x809[\x05EJoV\xB3\x90\x12\xE0^Bo\xB2~\x01S6\x0C\x8A\x94\xDE\xA4\x96\x91Kd.{\xB4\xB5S\xAAmP9\xD3\x9B\xEB\x8C\xC3\xDD\x14\xB1K5\xBA\x93\xCA\x17\xED\xB3@\xF10\xE0\xE4i\xC0\x95\x8F\x03\xCE>\x0FX\xDF\x88\xC67b\xE1\x1BQ|#\x96\xBE\x11\x8DoD\xEB\x1B\xB1\xF03\xE5\x1A\x0B\xBCh\x1E\\\x9Do\xA4\x86C\xC0\xCD*\x95\x8F\xF8\x1D\x09\xDF\x7Fn?\x04\x1A
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001160 00000 n 
0000000015 00000 n 
0000001248 00000 n 
0000001003 00000 n 
0000001314 00000 n 
0000001359 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 